core/src/main/java/io/grpc/ClientInterceptors.java

```diff
@@ -204,4 +218,195 @@ public final void start(Listener<RespT> responseListener, Metadata headers) {
+      }
+
+      @Override
+      public void start(Listener<RespT> listener, Metadata headers) {
+        checkState(attemptListeners.isEmpty());
+        checkState(responseListener == null);
+        checkState(requestHeaders == null);
+        responseListener = listener;
+        requestHeaders = headers;
+        ClientCall<ReqT, RespT> firstCall = channel.newCall(method, callOptions);
+        AttemptListener attemptListener = new AttemptListener(firstCall);
+        attemptListeners.add(attemptListener);
+        firstCall.start(attemptListener, headers);
+      }
+
+      @Override
+      public void request(int numMessages) {
+        lastCall().request(numMessages);
+      }
+
+      @Override
+      public void cancel() {
+        for (AttemptListener attempt : attemptListeners) {
+          attempt.call.cancel();
+        }
+      }
+
+      @Override
+      public void halfClose() {
+        lastCall().halfClose();
+      }
+
+      @Override
+      public void sendMessage(ReqT message) {
+        checkState(requestMessage == null);
+        requestMessage = message;
+        lastCall().sendMessage(message);
+      }
+
+      @Override
+      public boolean isReady() {
+        return lastCall().isReady();
+      }
+
+      @Override
+      public void setMessageCompression(boolean enabled) {
+        compressionEnabled = enabled;
+        lastCall().setMessageCompression(enabled);
+      }
+
+      private void maybeRetry(AttemptListener attempt) {
+        Status status = attempt.responseStatus;
+        if (!retryPolicy.isRetryable(status, method, callOptions)) {
+          return;
+        }
+
+        if (status.isOk()) {
+          useResponse(attempt);
+          return;
+        }
+
+        Status.Code code = status.getCode();
+        if (code == Status.Code.CANCELLED || code == Status.Code.DEADLINE_EXCEEDED) {
+          AttemptListener latest = latestResponse;
+          if (latest != null) {
+            useResponse(latest);
+          } else {
+            useResponse(attempt);
+          }
+          return;
+        }
+
+        ScheduledFuture<?> retryTask = scheduledExecutor.schedule(new Runnable() {
+          @Override
+          public void run() {
+            ClientCall<ReqT, RespT> nextCall = channel.newCall(method, callOptions);
+            AttemptListener nextAttemptListener = new AttemptListener(nextCall);
+            attemptListeners.add(nextAttemptListener);
+            nextCall.start(nextAttemptListener, requestHeaders);
+            nextCall.setMessageCompression(compressionEnabled);
+            nextCall.sendMessage(requestMessage);
+            // TODO(lukaszx0): In other places, we're preemptively fetching two requests for
+            // unary calls should we do the same here?
+            nextCall.request(1);
+            nextCall.halfClose();
+          }
+        }, retryPolicy.getNextBackoffMillis(), TimeUnit.MILLISECONDS);
```

Comment:
It would be nice to verify that there is time remaining in the call's deadline first. Also, you might want to save a reference to this future in a field so that you can cancel the scheduled task in `onCancel`.